src/coreclr/jit/flowgraph.cpp

```diff
@@ -4083,3 +4083,560 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
+            if (otherLoop->ContainsBlock(header))
+            {
+                // Ancestor loop; should contain all blocks of this loop
+                //
+                loop->VisitLoopBlocks([otherLoop](BasicBlock* loopBlock) {
+                    assert(otherLoop->ContainsBlock(loopBlock));
+                    return BasicBlockVisit::Continue;
+                });
+            }
+            else
+            {
+                // Non-ancestor loop; should have no blocks in common with current loop
+                //
+                loop->VisitLoopBlocks([otherLoop](BasicBlock* loopBlock) {
+                    assert(!otherLoop->ContainsBlock(loopBlock));
+                    return BasicBlockVisit::Continue;
+                });
+            }
+        }
+#endif
+
+        // Record this loop
+        //
+        loop->m_index = (unsigned)loops->m_loops.size();
+        loops->m_loops.push_back(loop);
+
+        JITDUMP("Added loop " FMT_LP " with header " FMT_BB "\n", loop->GetIndex(), loop->GetHeader()->bbNum);
+    }
+
+    if (loops->m_loops.size() > 0)
+    {
+        JITDUMP("\nFound %zu loops\n", loops->m_loops.size());
+    }
+
+    if (loops->m_improperLoopHeaders > 0)
+    {
+        JITDUMP("Rejected %u loop headers\n", loops->m_improperLoopHeaders);
+    }
+
+    return loops;
+}
+
+//------------------------------------------------------------------------
+// FlowGraphNaturalLoops::FindNaturalLoopBlocks: Find the loop blocks for a
+// specified loop.
+//
+// Parameters:
+//   loop     - The natural loop we are constructing
+//   worklist - Scratch worklist to use for the search
+//
+// Returns:
+//   True if the loop is natural; marks the loop blocks into 'loop' as part of
+//   the search.
+//
+bool FlowGraphNaturalLoops::FindNaturalLoopBlocks(FlowGraphNaturalLoop* loop, jitstd::list<BasicBlock*>& worklist)
+{
+    const FlowGraphDfsTree* tree       = loop->m_tree;
+    Compiler*               comp       = tree->GetCompiler();
+    BitVecTraits            loopTraits = loop->LoopBlockTraits();
+    BitVecOps::AddElemD(&loopTraits, loop->m_blocks, 0);
+
+    // Seed the worklist
+    //
+    worklist.clear();
+    for (FlowEdge* backEdge : loop->m_backEdges)
+    {
+        BasicBlock* const backEdgeSource = backEdge->getSourceBlock();
+        if (backEdgeSource == loop->GetHeader())
+        {
+            continue;
+        }
+
+        assert(!BitVecOps::IsMember(&loopTraits, loop->m_blocks, loop->LoopBlockBitVecIndex(backEdgeSource)));
+        worklist.push_back(backEdgeSource);
+        BitVecOps::AddElemD(&loopTraits, loop->m_blocks, loop->LoopBlockBitVecIndex(backEdgeSource));
+    }
+
+    // Work back through flow to loop head or to another pred
+    // that is clearly outside the loop.
+    //
+    while (!worklist.empty())
+    {
+        BasicBlock* const loopBlock = worklist.back();
+        worklist.pop_back();
+
+        for (FlowEdge* predEdge = comp->BlockPredsWithEH(loopBlock); predEdge != nullptr;
```

Comment:
This one is the important one to switch to `BlockPredsWithEH` to recognize the full SCC of loops containing EH constructs.